Feature/verify purchase

bizz84 opened the pull request.

bizz84 commented on SwiftyStoreKit/InAppReceipt.swift:
```diff
@@ -235,5 +241,69 @@ internal class InAppReceipt {
+            // https://developer.apple.com/library/mac/releasenotes/General/ValidateAppStoreReceipt/Chapters/ReceiptFields.html
+              let expires_date = receipt["expires_date_ms"] as? NSString
+              let expires_date_double = (expires_date?.doubleValue ?? 0.0) / 1000
+              return NSDate(timeIntervalSince1970: expires_date_double)
```
I'm not sure this code works as there isn't an `expires_date_ms` field [here](https://developer.apple.com/library/mac/releasenotes/General/ValidateAppStoreReceipt/Chapters/ReceiptFields.html).
In my tests I have added a consumable IAP and I have observed that after purchasing it successfully and calling the verifyPurchase() method, this call will return `NSDate(timeIntervalSince1970: 0)` which is _1 January 1970_ and means that the receipt is always expired.
By reading the docs it seems that if the `expires_date` or `expiration_date` fields are missing, the IAP can't expire and we should return `.Purchased(expiresDate: nil)`.
Instead, if we have an expiration date, we should compare it with the current date, and return `Purchased` or `Expired` based on that.
What do you think?

NicolasMahe replied:
The expires_date related fiels are only present for Automatically Renewable Subscription.
That's why the parameter `validUntil`is optional:
- If you want to test a consumable, non-consumable or non-renewable subscription, you should let the `validUntil` parameter to `nil`. In those case, the function return `.Purchased(expiresDate: nil)`.
- If you indicate a `validUntil`parameter to test a Automatically Renewable Subscription, the function will continue after the line 278 and test the `expires_date` with the `validUntil`.

I think the user's API should be split in 2 to avoid confusion:
- One only for Automatically Renewable Subscription with a required validUntil date.
- One for the others type of product with no validUntil parameter.
Those 2 API will use the same function in the background.


That's true that `expires_date_ms` is not in the docs, but it is present in the receipt json, as well as 'purchase_date_ms', 'original_purchase_date_ms', 'expires_date_pst', `original_purchase_date_pst` and `purchase_date_pst'. I think they didn't put them in the doc because they are just helper to parse the plain string date easier.
We can parse `expires_date` with a `NSDateFormatter` if you think it's better to use only fields present in the docs.

bizz84 replied:
I have updated the API so that it takes a `PurchaseType` enum as a parameter and this is used to determine whether the expiration date should be considered. This puts the responsibility on the client to decide which is the correct PurchaseType for each IAP, which I believe is a reasonable request. Should the client specify a `AutomaticallyRenewableSubscription` with a date for a purchase that really is `Consumable`, the completion block will likely say that the purchase is expired, however this can be classified as incorrect usage of the API.

bizz84 commented on SwiftyStoreKit/SwiftyStoreKit.swift:
```diff
@@ -158,7 +158,26 @@ public class SwiftyStoreKit {
+        InAppReceipt.verify(receiptVerifyURL: url, password: password, session: session) { result in
+         
+            dispatch_async(dispatch_get_main_queue()) {
+                completion(result: result)
```
This change is technically not part of this PR but I realised it's a fix that is needed otherwise we the completion block is not called on the main thread.

bizz84 commented on SwiftyStoreOSXDemo/ViewController.swift:
```diff
@@ -32,28 +32,29 @@ class ViewController: NSViewController {
-    @IBAction func getInfo2(sender: AnyObject!) {
-        getInfo("2")
+    @IBAction func getInfo2(sender: AnyObject?) {
+        getInfo("consumablePurchase")
```
Registered a new `consumablePurchase` ID so that we can test consumable IAPs.

bizz84 commented on SwiftyStoreKit/SwiftyStoreKit.swift:
```diff
@@ -158,7 +158,26 @@ public class SwiftyStoreKit {
+    public class func verifyPurchase(
+        productId productId: String,
+        inReceipt receipt: ReceiptInfo,
+        validUntil: NSDate? = nil
```
Shouldn't validUntil be non optional and default to `NSDate()`?

NicolasMahe replied:
See my previous response.

bizz84 commented on SwiftyStoreKit/InAppReceipt.swift:
```diff
@@ -30,10 +30,28 @@ public typealias ReceiptInfo = [String: AnyObject]
+        case AutomaticallyRenewableSubscription(validUntilDate: NSDate)
+        // Free subscriptions don’t expire and can only be offered in apps that are in the Magazines & Newspapers category.
+        case FreeSubscription
+        // Non-Renewing Subscriptions allow the sale of services with a limited duration.
```
Not sure if this should also have a `validUntilDate` parameter?

NicolasMahe replied:
Yes it should. The AutomaticallyRenewableSubscription need to be test with a date. If not, you will considered a expired subscription as valid.
By default the date should be `now`.

bizz84 commented on SwiftyStoreKit/InAppReceipt.swift:
```diff
@@ -30,10 +30,28 @@ public typealias ReceiptInfo = [String: AnyObject]
+        case Purchased(expiresDate: NSDate?)
+        case Expired(expiresDate: NSDate)
+        case NotPurchased
+    }
```
Should we introduce separate result enums depending on whether we are dealing with Consumable/NonConsumable or Subscriptions?

NicolasMahe replied:
I think it's a good idea. The API will be even more understandable.

bizz84 commented on SwiftyStoreKit/InAppReceipt.swift:
```diff
@@ -235,5 +253,73 @@ internal class InAppReceipt {
+            .map { (receipt) -> NSDate in
+                let expires_date = receipt["expires_date_ms"] as? NSString
+                let expires_date_double = (expires_date?.doubleValue ?? 0.0) / 1000
+                return NSDate(timeIntervalSince1970: expires_date_double)
```
Is there a way that we can more gracefully handle the case where the `expires_date` value is missing and not return "1 January 1970"?

NicolasMahe replied:
Yes, we can set the returned NSDate to an optional and check that expires_date is not missing:
```swift
let expiresDates = receiptsMatchingProductId
    .map { (receipt) -> NSDate? in
        guard let expires_date = receipt["expires_date_ms"] as? NSString
            else { return nil }
        let expires_date_double = expires_date.doubleValue / 1000
        return NSDate(timeIntervalSince1970: expires_date_double)
    }
```
We should also spread the consequences of this change to the end of the function.

bizz84 commented:
@NicolasMahe See updated implementation which introduces a `PurchaseType` type to deal with different types of IAPs.
One thing I noticed is that when I run the `verifyReceipt()` method I get a `ReceiptStatus.SecretNotMatching` error which maps tho this:
```
21004
The shared secret you provided does not match the shared secret on file for your account.
Only returned for iOS 6 style transaction receipts for auto-renewable subscriptions.
```
Another possible problem is that when I restore the purchases I get multiple transactions for my auto renewable subscriptions - not sure if this is expected or not.
Have you done some testing around Subscriptions? I would like to include the `verifyPurchase()` feature with support for subscriptions however I need to understand a bit better how this works if I'm going to maintain this code going forward.

NicolasMahe commented:
@bizz84 You need to indicate your Shared Secret from the iTunes Connect in-App Purchases page when you call the function `SwiftyStoreKit.verifyReceipt` in order to get the auto-renewable subscription receipt.

---

I have the same behaviour concerning the restore of auto-renewable subscriptions. When you subscribe to an auto-renewable subscription in the test endpoint environment, it is automatically renew maximum 6 times. The duration of the subscription is also different. For example, if you test a 1 month subscription, the test duration is only 5 mins. So if you restore the purchases 1h after the purchase of the first subscription, you will automatically have 5 new purchases. But the last one is already expired.
Apple established a system like these because you can't manage the subscriptions in the test environment. So these way, you can test the renew of a subscription but also its expiration.
[Apple docs](https://developer.apple.com/library/ios/documentation/LanguagesUtilities/Conceptual/iTunesConnectInAppPurchase_Guide/Chapters/TestingInAppPurchases.html)

NicolasMahe commented:
@bizz84 is there a way for me to push commit to these branch to help you improve the functionality ?

bizz84 commented:
@NicolasMahe you could try to fork this PR so that we can merge your changes here first, and later to master when we're happy that it all works.

bizz84 commented on SwiftyStoreKit/InAppReceipt.swift:
```diff
@@ -235,5 +248,101 @@ internal class InAppReceipt {
+            .sort { (a, b) -> Bool in
+                // Sort by descending date order
+                return a.compare(b) == .OrderedDescending
+            }
```
I have simplified this logic so that less map/filter methods are needed

bizz84 commented on SwiftyStoreKit/InAppReceipt.swift:
```diff
@@ -235,5 +248,101 @@ internal class InAppReceipt {
+      
+        guard let firstExpiryDate = expiryDateValues.first else {
+            return .NotPurchased
+        }
```
As we sort in the step above, the `firstExpiryDate` is the farthest in the future.

bizz84 commented on SwiftyStoreKit/InAppReceipt.swift:
```diff
@@ -235,5 +248,101 @@ internal class InAppReceipt {
+        }
+        else {
+            // The subscription is expired
+            return .Expired(expiresDate: firstExpiryDate)
```
If `firstExpiryDate` is expired, any other `expiryDates` will be as well due to `.OrderDescending`

bizz84 commented:
@NicolasMahe I have merged your PR and simplified some of the expiry dates logic. Please review and let me know if this makes sense.
Also updated README and aligned OSX and iOS demo code.

NicolasMahe commented:
Everything is ok for me 👍 
Nice modifications!
